<a href="https://colab.research.google.com/github/Farhan-Kebab/Pemdas_quest_TKJ4/blob/main/P02_Praktikum_RafiAfrand_Colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Praktikum Pertemuan 2 — pandas di Google Colab
**Mata Kuliah:** Exploratory Data Analysis (KMTI21133) — Pertemuan 2
**Sub-CPMK2:** Mahasiswa mampu menggunakan Python (pandas) di Google Colab/Jupyter dan Git/GitHub untuk memuat, menelusuri, dan menyimpan data serta kode analisis.

**Cara kerja:** sel bertanda `# DEMO` diisi bersama dosen; sel bertanda `# TODO` Anda kerjakan sendiri. Setiap TODO punya *expected output* agar Anda bisa memeriksa jawaban sendiri. Simpan notebook ini ke repositori tim di akhir sesi (Bagian E).

## Bagian 0 — Persiapan
Jalankan sel di bawah. Jika file `penjualan_umkm_2024.csv` belum ada di sesi Colab, sel akan meminta Anda mengunggahnya (ikon folder di kiri → Upload).

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', 30)
pd.set_option('display.width', 120)
print('pandas', pd.__version__, '| numpy', np.__version__)

pandas 2.2.3 | numpy 2.1.3


In [2]:
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/EDA02/Data/penjualan_umkm_2024.csv')

---
## Bagian A — Memuat dan memeriksa data (Latihan 1)
**Tujuan:** sebelum menganalisis apa pun, jawab empat pertanyaan: *berapa besar data ini, apa saja kolomnya, apa tipe datanya, dan seberapa bersih?*

In [4]:
# DEMO A1 — memuat data
df.head()

,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
0,TRX2024198,2024-11-14,Jakarta Selatan,Makanan,Sambal Roa,8,38000,5,Transfer,4.7,288800
1,TRX2025061,2024-08-16,Depok,Fashion,Kaos Lokal,8,131500,15,Tunai,3.8,894200
2,TRX2024249,2024-12-21,Depok,Fashion,Hijab Motif,12,102500,10,E-Wallet,4.3,1107000
3,TRX2024903,2024-09-27,Bekasi,Fashion,Kemeja Tenun,11,113500,10,E-Wallet,NaN,1123650
4,TRX2024097,2024-05-06,Depok,Kerajinan,Batik Tulis,8,123000,20,QRIS,4.5,787200


In [5]:
# DEMO A2 — ukuran dan struktur
print('Baris, kolom:', df.shape)
df.info()

Baris, kolom: (1212, 11)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1212 entries, 0 to 1211
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id_transaksi      1212 non-null   object 
 1   tanggal           1212 non-null   object 
 2   kota              1212 non-null   object 
 3   kategori          1212 non-null   object 
 4   produk            1212 non-null   object 
 5   jumlah            1212 non-null   int64  
 6   harga_satuan      1212 non-null   int64  
 7   diskon_persen     1212 non-null   int64  
 8   metode_bayar      1187 non-null   object 
 9   rating_pelanggan  1151 non-null   float64
 10  total_bayar       1212 non-null   int64  
dtypes: float64(1), int64(4), object(6)
memory usage: 104.3+ KB


In [6]:
# DEMO A3 — ringkasan numerik
df.describe()

,jumlah,harga_satuan,diskon_persen,rating_pelanggan,total_bayar
count,1212.000000,1212.000000,1212.000000,1151.000000,1.212000e+03
mean,7.548680,69210.808581,6.839934,4.152389,4.859965e+05
std,4.055776,50222.966019,7.446309,0.541938,4.826499e+05
min,1.000000,20000.000000,0.000000,2.500000,1.760000e+04
25%,4.000000,30500.000000,0.000000,3.800000,1.655000e+05
50%,8.000000,39500.000000,5.000000,4.200000,3.049750e+05
75%,11.000000,113000.000000,15.000000,4.600000,6.239625e+05
max,14.000000,191500.000000,20.000000,5.000000,2.667000e+06


> **Catatan dosen:** `describe()` secara default hanya meringkas kolom numerik. Coba `df.describe(include='object')` untuk kolom teks.

In [7]:
# TODO A4
# a. Tampilkan 5 baris TERAKHIR dari df
# b. Tampilkan daftar nama kolom beserta tipe datanya
# Expected: 1212 baris x 11 kolom; kolom 'tanggal' bertipe object (belum datetime)


In [8]:
df.tail()

,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
1207,TRX2025021,2024-11-18,Tangerang Selatan,Makanan,Keripik Singkong,1,22000,20,QRIS,4.1,17600
1208,TRX2024789,2024-09-16,Bogor,Minuman,Teh Herbal,5,55000,10,Transfer,NaN,247500
1209,TRX2024256,2024-08-13,Bogor,Minuman,Sirup Markisa,2,47000,10,QRIS,3.6,84600
1210,TRX2024968,2024-01-08,Bekasi,Kerajinan,Tas Anyaman,7,132000,20,Transfer,4.6,739200
1211,TRX2024952,2024-03-11,Bekasi,Minuman,Teh Herbal,7,29500,20,Transfer,4.4,165200


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1212 entries, 0 to 1211
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id_transaksi      1212 non-null   object 
 1   tanggal           1212 non-null   object 
 2   kota              1212 non-null   object 
 3   kategori          1212 non-null   object 
 4   produk            1212 non-null   object 
 5   jumlah            1212 non-null   int64  
 6   harga_satuan      1212 non-null   int64  
 7   diskon_persen     1212 non-null   int64  
 8   metode_bayar      1187 non-null   object 
 9   rating_pelanggan  1151 non-null   float64
 10  total_bayar       1212 non-null   int64  
dtypes: float64(1), int64(4), object(6)
memory usage: 104.3+ KB


In [10]:
# TODO A5 — seberapa bersih data ini?
# a. Hitung jumlah nilai kosong (missing) per kolom
# b. Hitung jumlah baris duplikat
# c. Hitung berapa nilai unik pada kolom 'kota'
# Expected: rating_pelanggan & metode_bayar memiliki missing; ada 12 baris duplikat;
#           kolom 'kota' punya LEBIH dari 5 nilai unik -- mengapa? (diskusikan)


In [11]:
df.isna().sum()

,0
id_transaksi,0
tanggal,0
kota,0
kategori,0
produk,0
jumlah,0
harga_satuan,0
diskon_persen,0
metode_bayar,25
rating_pelanggan,61


In [12]:
df.duplicated().sum()

np.int64(12)

In [13]:
df['kota'].unique()

array(['Jakarta Selatan', 'Depok', 'Bekasi', 'Tangerang Selatan', 'DEPOK',
       'Bogor', 'BOGOR', 'JAKARTA SELATAN', 'TANGERANG SELATAN', 'BEKASI'],
      dtype=object)

**Diskusi (2 menit):** mengapa jumlah nilai unik `kota` lebih banyak daripada jumlah kota sebenarnya? Masalah seperti ini akan kita tangani secara sistematis pada pertemuan 4–5.

---
## Bagian B — Seleksi dan filter (Latihan 2, berpasangan)
**Tujuan:** mengambil bagian data yang relevan dengan pertanyaan analisis.

In [14]:
# DEMO B1 — memilih kolom
df['total_bayar'].head()          # satu kolom -> Series
df[['kota','kategori','total_bayar']].head()   # beberapa kolom -> DataFrame

,kota,kategori,total_bayar
0,Jakarta Selatan,Makanan,288800
1,Depok,Fashion,894200
2,Depok,Fashion,1107000
3,Bekasi,Fashion,1123650
4,Depok,Kerajinan,787200


In [15]:
# DEMO B2 — loc (berbasis label) vs iloc (berbasis posisi)
#print(df.loc[0, 'produk'])
#print(df.iloc[0, 4])
df.loc[0:4, ['produk','jumlah','total_bayar']]

,produk,jumlah,total_bayar
0,Sambal Roa,8,288800
1,Kaos Lokal,8,894200
2,Hijab Motif,12,1107000
3,Kemeja Tenun,11,1123650
4,Batik Tulis,8,787200


In [16]:
# DEMO B3 — filter dengan kondisi (boolean mask)
mask = df['total_bayar'] > 1_000_000
print('Transaksi di atas 1 juta:', mask.sum())
df[mask].head()

Transaksi di atas 1 juta: 180


,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
2,TRX2024249,2024-12-21,Depok,Fashion,Hijab Motif,12,102500,10,E-Wallet,4.3,1107000
3,TRX2024903,2024-09-27,Bekasi,Fashion,Kemeja Tenun,11,113500,10,E-Wallet,NaN,1123650
36,TRX2024121,2024-05-31,Bekasi,Kerajinan,Batik Tulis,13,132000,5,Tunai,4.4,1630200
39,TRX2024692,2024-08-07,Depok,Kerajinan,Gerabah Hias,10,107000,0,Transfer,4.4,1070000
42,TRX2024232,2024-08-26,Tangerang Selatan,Kerajinan,Tas Anyaman,14,135500,0,QRIS,3.6,1897000


In [17]:
# DEMO B4 — menggabungkan kondisi
hi = df[(df['kategori'] == 'Makanan') & (df['diskon_persen'] >= 10)]
print(hi.shape)
df[df['metode_bayar'].isin(['QRIS','E-Wallet'])].shape

(201, 11)


(648, 11)

In [18]:
# TODO B5
# a. Ambil semua transaksi dengan rating_pelanggan < 3
# b. Berapa banyak transaksi tersebut dan berapa rata-rata total_bayar-nya?
# c. Ambil transaksi kategori 'Kerajinan' ATAU 'Fashion' dengan jumlah >= 10


In [19]:
#a
rating__rendah = df[df['rating_pelanggan']<3]
rating__rendah

,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar
25,TRX2025090,2024-11-27,Tangerang Selatan,Minuman,Sirup Markisa,6,55500,10,QRIS,2.8,299700
148,TRX2024429,2024-06-08,Jakarta Selatan,Fashion,Kaos Lokal,2,100500,0,Tunai,2.9,201000
231,TRX2024955,2024-01-30,Bogor,Makanan,Sambal Roa,12,26000,10,E-Wallet,2.5,280800
244,TRX2024349,2024-10-11,Depok,Minuman,Sirup Markisa,8,44500,0,Transfer,2.8,356000
361,TRX2024423,2024-09-05,Bogor,Minuman,Teh Herbal,12,46000,10,E-Wallet,2.9,496800
368,TRX2024214,2024-05-07,Depok,Minuman,Teh Herbal,4,48500,0,Tunai,2.8,194000
383,TRX2024188,2024-01-22,Depok,Fashion,Hijab Motif,6,118500,0,Tunai,2.9,711000
455,TRX2024886,2024-03-24,Jakarta Selatan,Makanan,Kue Kering,7,26000,5,E-Wallet,2.9,172900
471,TRX2024229,2024-11-03,Depok,Kerajinan,Batik Tulis,6,176000,0,Transfer,2.9,1056000
477,TRX2024751,2024-07-26,Tangerang Selatan,Minuman,Sirup Markisa,11,32500,15,QRIS,2.9,303875


In [20]:
#b
print("Total transaksi dengan rating dibawah 3 adalah", rating__rendah.shape[0])

Total transaksi dengan rating dibawah 3 adalah 19


In [21]:
#b
print('Total rata rata total_bayar dengan rating dibawah 3 adalah', rating__rendah['total_bayar'].mean())

Total rata rata total_bayar dengan rating dibawah 3 adalah 386563.15789473685


In [22]:
kerajinan_fashion_diatas_sepuluh = df[(df['kategori'] == 'Fashion') | (df['kategori'] == 'Kerajinan') & (df['jumlah'] >= 10)]
print(kerajinan_fashion_diatas_sepuluh.shape[0])

296


In [23]:
# TODO B6 — pertanyaan bisnis
# 'Transaksi tunai bernilai besar (> 500.000) di kota mana saja?'
# Tampilkan kolom kota, produk, jumlah, total_bayar, diurutkan dari total_bayar terbesar.
# Petunjuk: gunakan .sort_values(..., ascending=False) dan .head(10)


In [24]:
transaksidiatas500k = df[df['total_bayar'] > 500000]
transaksidiatas500k['kota'].unique().sum()

'DepokBekasiBogorTangerang SelatanJakarta SelatanBEKASIJAKARTA SELATANDEPOK'

In [25]:
totalbayarkotaprodukjumlah = df[['kota','produk','jumlah','total_bayar']]
totalbayarkotaprodukjumlah.sort_values(by='total_bayar', ascending=False).head(10)


,kota,produk,jumlah,total_bayar
385,Bekasi,Batik Tulis,14,2667000
97,Bekasi,Gerabah Hias,12,2286000
1062,Depok,Tas Anyaman,13,2249000
742,Jakarta Selatan,Batik Tulis,13,2242500
988,Bogor,Batik Tulis,14,2237200
698,Bogor,Tas Anyaman,14,2230200
488,Jakarta Selatan,Gerabah Hias,13,2164500
145,Bogor,Gerabah Hias,14,2163000
166,Bogor,Batik Tulis,12,2148000
398,Jakarta Selatan,Kemeja Tenun,14,2093000


---
## Bagian C — Kolom turunan dan pengurutan

In [26]:
# DEMO C1 — membuat kolom baru
df['harga_bruto'] = df['jumlah'] * df['harga_satuan']
df['hemat'] = df['harga_bruto'] - df['total_bayar']
df[['harga_bruto','total_bayar','hemat']].head()

,harga_bruto,total_bayar,hemat
0,304000,288800,15200
1,1052000,894200,157800
2,1230000,1107000,123000
3,1248500,1123650,124850
4,984000,787200,196800


In [27]:
# DEMO C2 — kolom tanggal
df['tanggal'] = pd.to_datetime(df['tanggal'])
df['bulan'] = df['tanggal'].dt.month
df['nama_hari'] = df['tanggal'].dt.day_name()
df[['tanggal','bulan','nama_hari']].head()

,tanggal,bulan,nama_hari
0,2024-11-14,11,Thursday
1,2024-08-16,8,Friday
2,2024-12-21,12,Saturday
3,2024-09-27,9,Friday
4,2024-05-06,5,Monday


In [28]:
# TODO C3
# a. Buat kolom 'kelas_nilai' berisi 'Besar' jika total_bayar > 500000, selain itu 'Kecil'
#    Petunjuk: np.where(kondisi, 'Besar', 'Kecil')
# b. Hitung berapa banyak transaksi pada tiap kelas (petunjuk: value_counts())
# c. Tampilkan 5 transaksi dengan penghematan (kolom 'hemat') terbesar


In [29]:
#a
df['kelas_nilai'] = np.where(df['total_bayar'] > 500000, 'Besar', 'Kecil')

In [30]:
#b
print(df['kelas_nilai'].value_counts().unique)

<bound method Series.unique of kelas_nilai
Kecil    845
Besar    367
Name: count, dtype: int64>


In [31]:
#c
df.sort_values(by='hemat', ascending=False).head()

,id_transaksi,tanggal,kota,kategori,produk,jumlah,harga_satuan,diskon_persen,metode_bayar,rating_pelanggan,total_bayar,harga_bruto,hemat,bulan,nama_hari,kelas_nilai
295,TRX2024224,2024-02-13,Jakarta Selatan,Kerajinan,Gerabah Hias,14,179500,20,Transfer,4.1,2010400,2513000,502600,2,Tuesday,Besar
155,TRX2024618,2024-06-15,Jakarta Selatan,Kerajinan,Batik Tulis,14,169500,20,E-Wallet,3.8,1898400,2373000,474600,6,Saturday,Besar
220,TRX2024902,2024-02-04,Jakarta Selatan,Kerajinan,Batik Tulis,13,182500,20,QRIS,4.1,1898000,2372500,474500,2,Sunday,Besar
112,TRX2025135,2024-07-05,Tangerang Selatan,Kerajinan,Gerabah Hias,13,176500,20,Transfer,4.0,1835600,2294500,458900,7,Friday,Besar
381,TRX2025146,2024-10-08,Bogor,Kerajinan,Gerabah Hias,13,163000,20,Transfer,NaN,1695200,2119000,423800,10,Tuesday,Besar


---
## Bagian D — Agregasi dengan groupby
Pola yang akan Anda pakai sepanjang semester: **split → apply → combine**.

In [32]:
# DEMO D1 — agregasi sederhana
df.groupby('kategori')['total_bayar'].mean().round(0)

,total_bayar
kategori,
Fashion,753346.0
Kerajinan,1043525.0
Makanan,209205.0
Minuman,291653.0


In [33]:
# DEMO D2 — beberapa agregasi sekaligus
ringkas = (df.groupby('kategori')
             .agg(jumlah_transaksi=('id_transaksi','count'),
                  total_penjualan=('total_bayar','sum'),
                  rata_rata_nilai=('total_bayar','mean'),
                  rata_rata_rating=('rating_pelanggan','mean'))
             .round(2)
             .sort_values('total_penjualan', ascending=False))
ringkas

,jumlah_transaksi,total_penjualan,rata_rata_nilai,rata_rata_rating
kategori,,,,
Kerajinan,244,254620050,1043524.80,4.14
Fashion,198,149162500,753345.96,4.12
Makanan,477,99790950,209205.35,4.19
Minuman,293,85454225,291652.65,4.13


In [34]:
# DEMO D3 — dua kunci pengelompokan
df.groupby(['kategori','metode_bayar'])['total_bayar'].sum().unstack(fill_value=0)

metode_bayar,E-Wallet,QRIS,Transfer,Tunai
kategori,,,,
Fashion,35309950,39074825,39080725,32852500
Kerajinan,48596275,94013100,50459425,57001000
Makanan,20332425,33687075,21838125,20904100
Minuman,12929075,32409575,23852550,15226000


In [35]:
# TODO D4
# a. Total penjualan per kota, diurutkan dari yang terbesar
# b. Rata-rata rating per kategori. Kategori mana yang ratingnya paling rendah?
# c. Penjualan per bulan. Bulan mana yang tertinggi?


In [36]:
#a
penjualanperkota = df.groupby('kota')['total_bayar'].sum().sort_values(ascending=False)
print(penjualanperkota)

kota
Depok                172719800
Jakarta Selatan      133467675
Bogor                105871075
Bekasi                95609225
Tangerang Selatan     65794250
JAKARTA SELATAN        5236300
BEKASI                 4139000
DEPOK                  3962400
BOGOR                  1136250
TANGERANG SELATAN      1091750
Name: total_bayar, dtype: int64


In [37]:
#b
ratingkategori = df.groupby('kategori')['rating_pelanggan'].mean().sort_values().head(1)
print(ratingkategori)

kategori
Fashion    4.122162
Name: rating_pelanggan, dtype: float64


In [38]:
#c
penjualanperbulan = df.groupby('bulan')['total_bayar'].sum().sort_values(ascending=False).head(2)
print(penjualanperbulan)

bulan
2    66337075
8    61894850
Name: total_bayar, dtype: int64


In [39]:
# TODO D5 — satu temuan
# Tulis SATU kalimat temuan dari hasil di atas dalam sel Markdown berikutnya.
# Format: 'X memiliki ... sebesar ..., lebih tinggi/rendah dibandingkan ...'
# Ingat: ini temuan DESKRIPTIF, belum boleh diklaim sebagai sebab-akibat.


**Temuan saya:** _(tulis di sini)_

Bulan Februari memiliki total_bayar terbesar sebesar 66337075, lebih tinggi dibandingkan bulan Agustus



---
## Bagian E — Menyimpan hasil dan mengunggah ke repositori tim

In [40]:
# DEMO E1 — menyimpan hasil agregasi
ringkas.to_csv('ringkasan_kategori.csv')
print('Tersimpan.')
# Di Colab: files.download('ringkasan_kategori.csv') untuk mengunduh

Tersimpan.


### Alur kerja repositori tim (dilakukan bersama di kelas)
```
nama-tim-eda/
├── README.md              <- tujuan proyek, anggota, cara menjalankan
├── requirements.txt
├── data/
│   ├── raw/               <- data asli, TIDAK PERNAH diubah
│   └── processed/         <- hasil olahan
├── notebooks/             <- 01_quality.ipynb, 02_cleaning.ipynb, ...
├── knime/
└── reports/
```
**Dua cara menyimpan notebook ke GitHub:**
1. Colab → *File → Save a copy in GitHub* → pilih repo tim → tulis pesan commit.
2. Melalui terminal/Git lokal:
```bash
git clone https://github.com/<organisasi-tim>/<nama-repo>.git
cd <nama-repo>
git checkout -b latihan-pert2
git add notebooks/P02_praktikum.ipynb
git commit -m 'Latihan pert 2: eksplorasi awal data penjualan UMKM'
git push origin latihan-pert2
```
**Aturan pesan commit di kelas ini:** kalimat perintah, maksimal ±60 karakter, menjelaskan *apa* dan *mengapa*.
Contoh baik: `Tambah agregasi penjualan per kota`. Contoh buruk: `update`, `fix`, `asdf`.

In [41]:
# TODO E2 — checklist sebelum commit
# 1. Restart runtime lalu Run all -> pastikan tidak ada error
# 2. Hapus sel percobaan yang tidak terpakai
# 3. Pastikan nama file mengikuti pola: P02_praktikum_<nama>.ipynb
# 4. Commit & push, lalu salin tautan commit Anda ke LMS


---
## Bagian F — Checkpoint proyek (dikerjakan bersama tim, di luar kelas)
Terapkan Bagian A pada **dataset calon proyek tim Anda**, lalu jawab di README repositori:
1. Berapa baris dan kolom? Apa satuan tiap barisnya (satu baris = satu apa)?
2. Kolom mana yang numerik, kategorikal, dan tanggal?
3. Kolom mana yang bermasalah (kosong, tipe salah, nilai tidak konsisten)?
4. Apakah data ini cukup untuk menjawab pertanyaan analisis tim? Jika belum, apa yang kurang?

Hasil poin 1–4 menjadi bahan **M1 (proposal proyek)** yang dikumpulkan pada pertemuan 3.